pkg/api/controller.go

```diff
@@ -5234,3 +5234,126 @@ func (c *Controller) GetUsageReportSummary(w http.ResponseWriter, r *http.Reques
 	}
 	writeResponse(w, r, http.StatusOK, response)
 }
+
+func (c *Controller) CreateUserExternalPrincipal(w http.ResponseWriter, r *http.Request, body apigen.CreateUserExternalPrincipalJSONRequestBody, userID, principalID string) {
+	ctx := r.Context()
+	if c.isExternalPrincipalNotSupported(ctx) {
+		writeError(w, r, http.StatusNotImplemented, "Not implemented")
+		return
+	}
+	if !c.authorize(w, r, permissions.Node{
+		Permission: permissions.Permission{
+			Action:   permissions.CreateUserExternalPrincipalAction,
+			Resource: permissions.UserArn(userID),
+		},
+	}) {
+		return
+	}
+
+	c.LogAction(ctx, "create_user_external_principal", r, "", "", "")
+
+	err := c.Auth.CreateUserExternalPrincipal(ctx, userID, principalID)
+	if c.handleAPIError(ctx, w, r, err) {
```

Comment:
You didn't add handling of `auth.NotImplemented` to `handleAPIError`, right?
Please add it there and also add tests to `Controller_test.go` to test that flow

Reply:
I'm confused and not sure why add it there? Also, how to handle it in that switch case `handleAPIError`? 

The reason for my confusin is: 

1. Mimicking the behavior of all other RBAC vs ACL, the API will return `http.StatusNotImplemented` before getting to that point. There is no path in the code that will pass that check and return `auth.NotImplemented`. 

2. If for some reason that will happened because of a code restructure or w/e then the error returned to the client will be `StatusInternalServerError` since it's really the case + `auth.NotImplemented` will be logged with `Error` log level in the server so administrators will see it.

Reply:
Oh I see what happened here. 
You check first thing for `isExternalPrincipalNotSupported` and then return not implemented. I thought we reach the `AuthService` that returns an error `auth.ErrNotImplemented` but I was wrong. When `handleAPIError` sees an error it doesn't recognize it will return `Internal Server Error` which is not what we want but also not the case here.

I believe it's slightly better to have the `AuthService` handle the possible `ExternalPrincipal`s enabled/disabled status in terms of separation of concerns, i.e. it's not the `controller` responsibility, you have errors as communication protocol between the 2 components. BUT - it is also not a bug, and definitely not a blocking bug.

Reply:
I prefer to keep it that way since all of our unsupported are on the controller level it's much easier to reason. 
In other words: standardization > normalization, it's easier to say "everything that is relying on RBAC auth service is on the controller level"